[Substrate.io Migration] - Polkadot Protocol > Protocol Components > Cryptography (#74)
polkadot-developers/polkadot-docs: polkadot-protocol/protocol-components/cryptography.md

---
title: Cryptography
description: A concise guide to cryptography in blockchain, covering hash functions, encryption types, digital signatures, and elliptic curve applications.
---

# Cryptography

## Introduction

> Cryptography forms the backbone of blockchain technology, providing the mathematical verifiability crucial for consensus systems, data integrity, and user security. While a deep understanding of the underlying mathematical processes isn't necessary for most blockchain developers, grasping the fundamental applications of cryptography is essential. This page offers a comprehensive overview of various cryptographic implementations used across Polkadot SDK and the broader blockchain ecosystem.

Reviewer: ```suggestion
Cryptography forms the backbone of blockchain technology, providing the mathematical verifiability crucial for consensus systems, data integrity, and user security. While a deep understanding of the underlying mathematical processes isn't necessary for most blockchain developers, grasping the fundamental applications of cryptography is essential. This page comprehensively overviews various cryptographic implementations used across Polkadot SDK-based chains and the broader blockchain ecosystem.
```

## Hash Functions

> Hash functions are fundamental to blockchain technology, creating a unique digital fingerprint for any piece of data including simple text, images, or any other form of file. They map input data of any size to a fixed-size output (typically 32 bytes) using complex mathematical operations. Hashing is used for verifying data integrity, creating digital signatures, and providing a secure way to store passwords. This form of mapping is known as the 'pigeonhole principle', and it is primarily implemented to efficiently and verifiably identify data from large sets.

Reviewer: I'd add a link to wikipedia or another source that explains what 'pigeonhole principle' is actually

### Key Properties of Hash Functions

- **Deterministic** - the same input always produces the same output
- **Quick computation** - it's easy to calculate the hash value for any given input
- **Pre-image resistance** - it's infeasible to generate the input data from its hash
- **Small changes in input yield large changes in output** - known as the avalanche effect

> - **Collision resistance** - it's extremely difficult to find two different inputs with the same hash

Reviewer: As these elements don't refer to UI elements, it'd be better to avoid bold font (maybe backticks?)

> - **Collision resistance** - it's extremely difficult to find two different inputs with the same hash

Reviewer: IMO, I'd suggest tweaking this line a bit. Instead of "extremely difficult," I'd say "the probabilities are extremely low." But the decision is up to you

#### Advantages

- Fast and efficient for large amounts of data
- Requires less computational power

#### Disadvantages

- Key distribution can be challenging
- Scalability issues in systems with many users

### Asymmetric Cryptography

Asymmetric encryption is a type of cryptography which uses two different keys, known as a keypair: a public key, used to encrypt plain text, and a private counterpart, used to decrypt the cipher text.

The public key is used to encrypt a fixed length message that can only be decrypted with the recipient's private key and, at times, a set password. The public key can be used to cryptographically verify that the corresponding private key was used to create a piece of data without compromising the private key itself, such as with digital signatures. This has obvious implications for identity, ownership and properties, and is used in many different protocols across both web2 and web3.

#### Advantages

- Solves the key distribution problem
- Enables digital signatures and secure key exchange

#### Disadvantages

- Slower than symmetric encryption
- Requires more computational resources

> ### Trade-offs and compromises

Reviewer: ```suggestion
### Trade-offs and Compromises
```

Symmetric cryptography is faster and requires fewer bits in the key to achieve the same level of security that asymmetric cryptography provides. However, it requires a shared secret before communication can take place, which poses issues to it's integrity and a potential compromise point. Asymmetric cryptography, on the other hand, does not require the secret to be shared ahead of time, allowing for far better end-user security.

Hybrid symmetric and asymmetric cryptography is often used to overcome the engineering issues of asymmetric cryptography, as it is slower and requires more bits in the key to achieve the same level of security. It is used to encrypt a key, and then use the comparatively lightweight symmetric cipher to do the 'heavy lifting' with the message.

> ## Digital signatures

Reviewer: ```suggestion
## Digital Signatures
```

1. The sender creates a hash of the message
2. The hash is encrypted using the sender's private key, creating the signature
3. The message and signature are sent to the recipient
4. The recipient decrypts the signature using the sender's public key
5. The recipient hashes the received message and compares it to the decrypted hash

If the hashes match, the signature is valid, confirming the message's integrity and the sender's identity

## Elliptic Curve

Blockchain technology requires the ability to have multiple keys creating a signature for block proposal and validation. To this end, Elliptic Curve Digital Signature Algorithm (ECDSA) and Schnorr signatures are two of the most commonly used methods. While ECDSA are a far simpler implementation, Schnorr signatures are more efficient when it comes to multi-signatures.

Schnorr signatures bring some noticeable features over the ECDSA/EdDSA schemes:

- It is better for hierarchical deterministic key derivations
- It allows for native multi-signature through [signature aggregation](https://bitcoincore.org/en/2017/03/23/schnorr-signature-aggregation/){target=\_blank}
- It is generally more resistant to misuse

> One sacrifice that is made when using Schnorr signatures over ECDSA is that both require 64 bytes, but only ECDSA signatures communicate their public key.

Reviewer: Nit picking here, but I'd briefly explain why the requirements of 64 bytes is a sacrifice. Probably, you can add something like:

"One sacrifice when using Schnorr signatures over ECDSA is that both require 64 bytes. This can potentially increase the total communication overhead, but only ECDSA signatures communicate their public key."

> ### Various implementations

Reviewer: ```suggestion
### Various Implementations
```
